R/ResampleInstances.R

```diff
@@ -67,3 +67,78 @@ instantiateResampleInstance.SpRepCVDesc = function(desc, size, task = NULL, coor
   g = as.factor(rep(seq_len(desc$reps), each = folds))
   makeResampleInstanceInternal(desc, size, train.inds = train.inds, test.inds = test.inds, group = g)
 }
+
+
+instantiateResampleInstance.FixedCVDesc = function(desc, size, task = NULL, coords) {
+  initial.window.abs = floor(desc$initial.window * size)
+  assertInt(initial.window.abs, lower = 1)
+  if (size - initial.window.abs < desc$horizon)
+    stop(catf("The initial window is %i observations while the data is %i observations. \n
+              There is not enough data left (%i observations) to create a test set for a %i size horizon",
+              initial.window.abs, size, initial.window.abs - size, desc$horizon))
+  skip = desc$skip
+  stops  = (seq(size))[initial.window.abs:(size - desc$horizon)]
+  starts = stops - initial.window.abs + 1
+  train.inds = mapply(seq, starts, stops, SIMPLIFY = FALSE)
+  test.inds  = mapply(seq, stops + 1, stops + desc$horizon, SIMPLIFY = FALSE)
+
+  thin = function(x, skip = 0) {
+    n = length(x)
+    x[seq(1, n, by = skip)]
+  }
+
+  if (skip > 0) {
+    train.inds = thin(train.inds, skip = skip)
+    test.inds = thin(test.inds, skip = skip)
+  }
+
+  if (length(test.inds) == 0)
+    stop("Skip is too large and has removed all resampling instances. Please lower the value of skip.")
+
+  # If the last value if not included, shift everything over by one
```

Comment:
What exactly is this supposed to do? Make sure that the last observation is in there? Wouldn't that require adjusting skip/horizon?

Reply:
It's there to ensure that last observation is included. There are instances where an initial window + horizon + skip can cause the most recent observation to not be included. This was a simple thing to say, "Hey if the most recent observation did not make it into your final iteration then we are going to shift everything up by one." Though this does mean you lose the first observation.

Do you think it would be better to do something that shifts everything forward while keeping that first observation (like below) or should I do nothing at all (maybe a warning?)

```R
train.inds = lapply(train.inds, function(x) c(x, x[length(x)] + 1))
test.inds = lapply(test.inds, function(x) c(x[2:desc$horizon], x[desc$horizon] + 1
```

Reply:
Hmm, I would just output a warning here without adjusting things manually -- maybe this is what the user intended? Certainly changing stuff internally without telling the user isn't good.

Is there some way we could make it easier for the user to specify these values? E.g. allow to specify skip/horizon in terms of the window size (if number >= 1 it's the number of observations, else it's the fraction of the window). Would that make sense?

Reply:
> Hmm, I would just output a warning here without adjusting things manually -- maybe this is what the user intended? Certainly changing stuff internally without telling the user isn't good.

Yes I agree that a warning would be best here

> Is there some way we could make it easier for the user to specify these values? E.g. allow to specify skip/horizon in terms of the window size (if number >= 1 it's the number of observations, else it's the fraction of the window). Would that make sense?

By these values do you mean skip/horizon? You usually want to forecast for a fixed length of time (say 10 days or so) so I'm not sure how setting these as a fraction of the window would feel. I can play around with setting it as a fraction of the window, but I think it's probably better to leave these as integers

Reply:
I've thought about this more and I think it's best to leave these as integers